src/libpmemfile/preload.c

```diff
@@ -2721,16 +2721,133 @@ establish_mount_points(const char *config)
 	if (getcwd(cwd, sizeof(cwd)) == NULL)
 		exit_with_msg(PMEMFILE_PRELOAD_EXIT_GETCWD_FAILED, "!getcwd");
 
-	struct stat kernel_cwd_stat;
-	if (stat(cwd, &kernel_cwd_stat) != 0) {
+	if (stat(cwd, kernel_cwd_stat) != 0) {
 		exit_with_msg(PMEMFILE_PRELOAD_EXIT_CWD_STAT_FAILED,
 				"!fstat cwd");
 	}
+}
 
-	assert(pool_count == 0);
+static void
+init_pool(struct pool_description *pool_desc, struct stat *kernel_cwd_stat)
+{
+	/* fetch pool_desc-fd, pool_desc->stat */
+	open_mount_point(pool_desc);
+
+	pool_desc->pool = NULL;
+
+	util_mutex_init(&pool_desc->pool_open_lock);
+	util_mutex_init(&pool_desc->process_switching_lock);
+
+	++pool_count;
+
+	/*
+	 * If the current working directory is a mount point, then
+	 * the corresponding pmemfile pool must opened at startup.
+	 * Normally, a pool is only opened the first time it is
+	 * accessed, but without doing this, the first access would
+	 * never be noticed.
+	 */
+	if (same_inode(&pool_desc->stat, kernel_cwd_stat)) {
+		open_new_pool(pool_desc);
+		if (pool_desc->pool == NULL) {
+			exit_with_msg(PMEMFILE_PRELOAD_EXIT_POOL_OPEN_FAILED,
+				"!opening pmemfile_pool");
+		}
+		cwd_pool = pool_desc;
+	}
+}
+
+static void
+detect_mount_points(struct stat *kernel_cwd_stat)
+{
+	FILE *file = fopen("/proc/self/mountinfo", "r");
 
+	if (!file)
+		return;
+
+	unsigned mount_id, parent_id, major, minor;
+	char root[PATH_MAX];
+	char mount_point[PATH_MAX];
+	char mount_options[4096];
+	char f[9][4096];
+	int matched = 0;
+	size_t len = PATH_MAX;
+	char *line = malloc(len);
```

Comment:
Is this needed when using `getline`?

Reply:
It's not. It's supposed to help limit the number of allocations.